### PR TITLE
refactor(web): OSK modularization - specialized OSKView classes ✂️

### PR DESCRIPTION
**Pull request by jahorton**
Follows #5620.

The main thing here:  formally splitting the old `OSKManager` into two specialized `OSKView` subclasses:
- `AnchoredOSKView`
    - Is "anchored" to an edge of the viewport
        -  For now... just the bottom edge.
        - Nomenclature (the "anchored" bit) follows standard UI layout conventions.
    - This inherits all of 14.0's touch-oriented OSK behavior.
- `FloatingOSKView`
    - 'Floats' about the page automatically unless the user specifies otherwise
    - This inherits all of 14.0's desktop OSK behavior.

While this certainly doesn't fully clean up a lot of the OSK's internal code messes, this does at least cleanly _separate_ those messes into their own areas.  There's certainly more clean-up work to be done, but there were already a lot of "low-hanging" distinctions that were easy to implement without much issue.

A prominent example of those low-hanging distinctions: many OSK API functions are far shorter in `AnchoredOSKView` because they simply never applied to touch keyboards.  (Their original versions explicitly early-`return`ed if a touch keyboard was active.)

On the other hand, there are a number of functions from the old `OSKManager` that need significant maintenance but require significant work... each.  So, for now those functions are preserved.  Many will likely remain for 15.0 due to the need to triage our remaining work items, though some will likely be addressed in upcoming work.

**Notes for _FUTURE_ work**:

- The "preserved" visibility fields and functions are a high-priority target for future work; there's a cobbled-together state machine in there that's quite poorly documented... and that wouldn't even apply to an "inlined" OSKView model.
    - "Inlined" - with positioning controlled like our keyboard documentation "renders", except actually being an _interactive_ keyboard, rather than static.
    - This might be a good model for the Developer keyboard test page; it'd keep the OSK locked in place when swapping among  possible target devices for a keyboard.  Hence the "high-priority."
    - That state machine also manages a "third" state:  there's a distinction between how embedded-touch is handled (where the keyboard is _always_ visible) vs how site-oriented touch is handled (where it is not).  The other state being how desktop OSKs are handled.

**Other Notes**:

Note:  when reviewing this PR, both `anchoredOskView.ts` and `floatingOskView.ts` should be compared against the old `oskManager.ts`, as both started as duplicates of that file.

Reviewers may find the differences between each pair of the two classes' API-oriented functions and their `_Show` function implementations to be quite interesting.  It's pretty clear that the API were only ever intended for desktop OSKs, so there's a lot of room for potential work in the future there.

# User Testing

@keymanapp/testers 

_Most_ of this PR is just shuttling code around and pruning dead code paths that resulted, so I don't expect much breakage compared to #5620.  But... this is a _very_ critical step that needs to be highly vetted; if I did something wrong here, that'll cause lots of problems later.

### SUITE_DESKTOP: `desktop` form factor tests

**Platform / browser combinations**:
- **GROUP_WIN**: Windows / Chrome
- **GROUP_MAC**: macOS / Firefox
    - Unless otherwise specified, use the "**Test unminified KeymanWeb**" test page.

**Suite tests**

- **TEST_RESIZING**:  Verify that desktop OSK resizing operates normally.

    <details><summary>Precise instructions</summary>

    Using the resize handle (image) for the desktop OSK, resize it at least three different ways.
     - The test "passes" if the OSK rescales according to the resize-handle's new position each time.  (The aspect ratio is _not_ locked, so narrow+tall and wide+short are both possible.)

    </details>

- **TEST_HARDWARE_SHIFT**:  A physical SHIFT keypress should affect the OSK's shift key.

    Using the `khmer_angkor` keyboard, press and hold your physical keyboard's `SHIFT` key.
    - Expected result:  on the OSK, the `shift` layer should be displayed and the `SHIFT` key highlighted.

- **TEST_OSK_KEY_SEQUENCE**:  Using the OSK with a keyboard for English, type "The ".  (Include the space)

    <details><summary>Detailed key sequence</summary>

    So, in sequence:
    - SHIFT
    - T
    - SHIFT (to return to the default layer)
    - h
    - e
    - (spacebar)

    </details>

- **TEST_OSK_HIDE**: click the 'x' at the OSK's top-right when displayed.

    The OSK should become hidden when you do so.  (It may be redisplayed through one of the UI options, often with a button or option with this icon:  image)

- **TEST_COOKIE_CLEAR**: tests default size for the OSK on a user's first visit to a page

    1. Open the test page
    2. Clear your browser's cache (usually an option within its settings menu)
    3. Reload the test page
    4. Select one of the "Type here" / "or here" controls in order to display the OSK.

    The OSK should have a "normal" size, rather than a "squashed" appearance.

    <details><summary>References</summary>

    - "Normal": image
        - The keys are nearly square here.
        - Something like the picture above would indicate a **PASS**.

    - "Squashed": image
        - The keys are far shorter than they are wide here.
        - Something like the picture above would indicate a **FAIL**.

    </details>

### **SUITE_PHONE**:  `phone` form-factor tests (both in-app and out-of-app)

**Platform / browser combinations**:
- **GROUP_CHROME_EMU**:  Chrome emulation / iOS 

    Use the "**Prediction - robust testing**" test page along with the the "Emulate a focused page" setting on the "Rendering" tab.

- **GROUP_ANDROID**:  Android / Keyman app

    Ensure that predictive text is enabled for English.

- **GROUP_IOS**:  iOS / System keyboard mode

    Ensure that predictive text is enabled for English.

**Suite tests**

- **TEST_ROTATE_P-TO-L**:

    1. With Keyman not active, load it in a portrait orientation.
    2. Once loaded, rotate the device to a landscape orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.

- **TEST_ROTATE_L-TO-P**:

    1. With Keyman not active, load it in a landscape orientation.
    2. Once loaded, rotate the device to a portrait orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.

- **TEST_LAYER_CAPTIONING**:  Touch-OSK layer shifting + spacebar captioning

    Using the `sil_euro_latin` keyboard, swap among the 'default', 'shift', 'numeric', 'symbol', and 'currency' layers.
    - Expected results:
        - the layer should always change correctly.
        - the space bar should be properly captioned on each layer.

- **TEST_SWAP**:  Keyboard swapping via globe key

    With at least two total keyboards available / installed, confirm that swapping between them works properly.

    </details>

### CODE REVIEW

**Comment by MakaraSok** [comment]
## SUITE_DESKTOP: `desktop` form factor tests
* GROUP_WIN: Windows / Chrome
  - **TEST_RESIZING: PASSED**
  - **TEST_HARDWARE_SHIFT: PASSED**
  - **TEST_OSK_KEY_SEQUENCE: PASSED**
  - **TEST_OSK_HIDE: PASSED** 
  Note that the OSK reappear to the top left corner of the screen instead of the previously set.
  image
  
  - **TEST_COOKIE_CLEAR: PASSED**

* GROUP_MAC: macOS / Firefox
  - **TEST_RESIZING: PASSED**
  - **TEST_HARDWARE_SHIFT: PASSED**
  - **TEST_OSK_KEY_SEQUENCE: PASSED**
  - **TEST_OSK_HIDE: PASSED**
  Note that the OSK reappear to the top left corner of the screen instead of the previously set.
  image

  - **TEST_COOKIE_CLEAR: PASSED**


## SUITE_PHONE: `phone` form-factor tests (both in-app and out-of-app)
* GROUP_CHROME_EMU: Chrome emulation / iOS 
  - **TEST_ROTATE_P-TO-L: FAILED** The OSK disappears after rotating.
  image
  - **TEST_ROTATE_L-TO-P: PASSED** The OSK rotates as expected.
  - **TEST_LAYER_CAPTIONING: PASSED**
  - **TEST_SWAP: PASSED**



* GROUP_IOS: iOS / System keyboard mode
  - **TEST_ROTATE_P-TO-L: PASSED** Seamlessly rotates. 
  - **TEST_ROTATE_L-TO-P: PASSED** Seamlessly rotates.
  - **TEST_LAYER_CAPTIONING: PASSED**
  - **TEST_SWAP: PASSED** The keyboard swaps properly, but 
  the spacebar caption is not visible on the numeric layer of Khmer Angkor.
  image

**Comment by MakaraSok** [comment]
## SUITE_PHONE: `phone` form-factor tests (both in-app and out-of-app)
* GROUP_ANDROID: Android / Keyman app
  - **TEST_ROTATE_P-TO-L: PASSED**
  - **TEST_ROTATE_L-TO-P: PASSED**
  - **TEST_LAYER_CAPTIONING: FAILED**
    On Khmer Angkor keyboard, the spacebar caption doesn't show on the numeric layer.
    

  - **TEST_SWAP: PASSED** The swap is not as smooth, but it might be by design. For what's worth, it should be as seamless as on iOS; however, it does switch to the next keyboard.

File: web/source/osk/oskView.ts [hunk 1]
```diff
@@ -205,7 +215,7 @@ namespace com.keyman.osk {
      * and the primary keyboard visualization elements.
      */
     get baseFontSize(): string {
-      return this.parsedBaseFontSize.styleString;
+      return this.parsedBaseFontSize?.styleString;
```
Comment (mcdurdin) [inline web/source/osk/oskView.ts:218]: Is this a change in result type that might cause issues elsewhere?

File: web/source/osk/oskView.ts [hunk 2]
```diff
@@ -545,5 +557,183 @@ namespace com.keyman.osk {
...
+     * Display build number
+     * 
+     * In the future, this should raise an event that the consuming KeymanWeb
+     * engine may listen for & respond to, rather than having it integrated
+     * as part of the OSK itself. 
+     */
+    showBuild() {
+      let keymanweb = com.keyman.singleton;
+      keymanweb.util.internalAlert('KeymanWeb Version '+keymanweb['version']+'.'+keymanweb['build']+'<br /><br />'
+          +'<span style="font-size:0.8em">Copyright &copy; 2017 SIL International</span>');
```
Comment (mcdurdin) [inline web/source/osk/oskView.ts:617]: Should probably fixup the copyright here!

**Review by mcdurdin — APPROVED** [review]
It's pretty hard to follow the differences between anchoredOskView and floatingOskView, so will trust to the user testing to pick up any discrepancies.

**Comment by jahorton** [comment]
> It's pretty hard to follow the differences between anchoredOskView and floatingOskView...

It always was.  That's why I've been pushing for this for so long; making them separate will make such differences a lot more obvious/distinct in the future.

> ... so will trust to the user testing to pick up any discrepancies.

👍

**Comment by jahorton** [comment]
Note for later:  will need to ensure https://github.com/keymanapp/keyman/pull/5619/commits/9a205474ccf3eeb5cc25399afe55c44509dfbb67 is properly adjusted here when merged.

**Comment by jahorton** [comment]
It looks like the issue with `khmer_angkor`'s spacebar caption on the numeric layer is on `master`, but not on `stable-14.0`.  It'll be addressed separately as #5688.

**Comment by jahorton** [comment]
The mobile-device bugs (aside from spacebar captions) should now be ready for a retest.

Spacebar captions are handled separately in #5688.

**Comment by jahorton** [comment]
Ah, right, for SUITE_PHONE / GROUP_CHROME_EMU / TEST_ROTATE_P-TO-L, you'll probably need the "Emulate a focused page" setting under the rendering tab in Chrome's developer mode.

Alternatively, running this test through an Android emulator connected to the TeamCity test page artifacts should also suffice.

**Comment by jahorton** [comment]
@keymanapp/testers This PR's failed tests should now be ready for re-testing.

Note: 

> Ah, right, for SUITE_PHONE / GROUP_CHROME_EMU / TEST_ROTATE_P-TO-L, you'll probably need the "Emulate a focused page" setting under the "Rendering" tab in Chrome's developer mode.
> 
> Alternatively, running this test through an Android emulator connected to the TeamCity test page artifacts should also suffice.

**Comment by jahorton** [comment]
Here goes something:

SUITE_PHONE
GROUP_CHROME_EMU

TEST_ROTATE_P-TO-L OPEN retest

GROUP_ANDROID

TEST_LAYER_CAPTIONING retest

**Comment by MakaraSok** [comment]
## SUITE_PHONE
  * GROUP_CHROME_EMU
    * **TEST_ROTATE_P-TO-L: PASSED** The OSK does rotate, and it works as expected.

  * GROUP_ANDROID
    * **TEST_LAYER_CAPTIONING: PASSED** The caption shown consistently on all layers in both `sil_euro_latin` and `khmer_angkor`.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.115-alpha](https://keyman.com/downloads/releases/alpha/15.0.115)
